torch/csrc/jit/interpreter.cpp

```diff
@@ -184,6 +185,22 @@ Operation createEvalOperation(CppOp * op) {
 }
 
 using tensor_list = std::vector<at::Tensor>;
+
+tensor_list unsafeToTensorListShare(const list_of_retainable & inputs) {
+  tensor_list tinputs;
+  tinputs.reserve(inputs.size());
+  for(auto & i : inputs) {
+    tinputs.push_back(unsafeToTensorShare(i));
+  }
+  return tinputs;
```

Comment:
nit: this is exactyl `fmap(inputs, unsafeToTensorShare)`

Reply:
I very frequently get these errors trying to use `fmap`, like in this case for this particular line. I don't feel confortable forcing other to maintain this code if the error messages for modifying it wrong are this bad.

```
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/local/cuda/include -I/data/users/zdevito/pytorch -I/data/users/zdevito/pytorch/torch/csrc -I/data/users/zdevito/pytorch/torch/lib/pybind11/include -I/data/users/zdevito/pytorch/torch/lib/tmp_install/include -I/data/users/zdevito/pytorch/torch/lib/tmp_install/include/TH -I/data/users/zdevito/pytorch/torch/lib/tmp_install/include/THNN -I/data/users/zdevito/pytorch/torch/lib/tmp_install/include/ATen -I/home/zdevito/local/anaconda2/lib/python2.7/site-packages/numpy/core/include -I/data/users/zdevito/pytorch/torch/lib/tmp_install/include/THD -I/usr/local/cuda/include -I/data/users/zdevito/pytorch/torch/lib/tmp_install/include/THCUNN -I/home/zdevito/local/anaconda2/include/python2.7 -c torch/csrc/jit/interpreter.cpp -o build/temp.linux-x86_64-2.7/torch/csrc/jit/interpreter.o -D_THP_CORE -std=c++11 -Wno-write-strings -fno-strict-aliasing -Wno-missing-braces -DWITH_NUMPY -DWITH_DISTRIBUTED -DWITH_CUDA -DCUDA_LIB_PATH=/usr/local/cuda/lib64 -DWITH_NCCL -DWITH_CUDNN -MMD -MF build/temp.linux-x86_64-2.7/torch/csrc/jit/interpreter.d
torch/csrc/jit/interpreter.cpp: In function ‘torch::jit::tensor_list torch::jit::unsafeToTensorListShare(const list_of_retainable&)’:
torch/csrc/jit/interpreter.cpp:196:42: error: no matching function for call to ‘fmap(const list_of_retainable&, at::Tensor (&)(at::Retainable*))’
   return fmap(inputs, unsafeToTensorShare);
                                          ^
torch/csrc/jit/interpreter.cpp:196:42: note: candidates are:
In file included from /data/users/zdevito/pytorch/torch/csrc/jit/tracer.h:6:0,
                 from /data/users/zdevito/pytorch/torch/csrc/autograd/function.h:15,
                 from /data/users/zdevito/pytorch/torch/csrc/jit/generated/aten_dispatch.h:2,
                 from torch/csrc/jit/interpreter.cpp:5:
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:13:23: note: template<class F, class T, class R> std::vector<R> torch::fmap(at::ArrayRef<TBase>, const F&)
 inline std::vector<R> fmap(at::ArrayRef<T> inputs, const F& fn) {
                       ^
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:13:23: note:   template argument deduction/substitution failed:
torch/csrc/jit/interpreter.cpp:196:42: note:   ‘std::vector<at::Retainable*>’ is not derived from ‘at::ArrayRef<TBase>’
   return fmap(inputs, unsafeToTensorShare);
                                          ^
In file included from /data/users/zdevito/pytorch/torch/csrc/jit/tracer.h:6:0,
                 from /data/users/zdevito/pytorch/torch/csrc/autograd/function.h:15,
                 from /data/users/zdevito/pytorch/torch/csrc/jit/generated/aten_dispatch.h:2,
                 from torch/csrc/jit/interpreter.cpp:5:
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:22:23: note: template<class F, class T, class R> std::vector<R> torch::fmap(const std::vector<T>&, const F&)
 inline std::vector<R> fmap(const std::vector<T>& inputs, const F& fn) {
                       ^
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:22:23: note:   template argument deduction/substitution failed:
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:21:34: error: function returning a function
 template<typename F, typename T, typename R = typename std::result_of<F(T)>::type>
                                  ^
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:27:23: note: template<class R, class T> std::vector<_RealType> torch::fmap(at::ArrayRef<TBase>)
 inline std::vector<R> fmap(at::ArrayRef<T> inputs) {
                       ^
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:27:23: note:   template argument deduction/substitution failed:
torch/csrc/jit/interpreter.cpp:196:42: note:   ‘std::vector<at::Retainable*>’ is not derived from ‘at::ArrayRef<TBase>’
   return fmap(inputs, unsafeToTensorShare);
                                          ^
In file included from /data/users/zdevito/pytorch/torch/csrc/jit/tracer.h:6:0,
                 from /data/users/zdevito/pytorch/torch/csrc/autograd/function.h:15,
                 from /data/users/zdevito/pytorch/torch/csrc/jit/generated/aten_dispatch.h:2,
                 from torch/csrc/jit/interpreter.cpp:5:
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:37:23: note: template<class R, class T> std::vector<_RealType> torch::fmap(const std::vector<T>&)
 inline std::vector<R> fmap(const std::vector<T>& inputs) {
                       ^
/data/users/zdevito/pytorch/torch/csrc/utils/functional.h:37:23: note:   template argument deduction/substitution failed:
torch/csrc/jit/interpreter.cpp:196:42: note:   candidate expects 1 argument, 2 provided
   return fmap(inputs, unsafeToTensorShare);
                                          ^
torch/csrc/jit/interpreter.cpp:197:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
[3/3] gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/local/cuda/include -I/data/users/zdevito/pytorch -I/data/users/zdevito...MPY -DWITH_DISTRIBUTED -DWITH_CUDA -DCUDA_LIB_PATH=/usr/local/cuda/lib64 -DWITH_NCCL -DWITH_CUDNN -MMD -MF build/temp.linux-x86_64-2.7/torch/csrc/jit/test_jit.d
ninja: build stopped: subcommand failed.
Command '['/home/zdevito/local/anaconda2/lib/python2.7/site-packages/ninja/data/bin/ninja', '-f', 'build/build.torch._C.ninja']' returned non-zero exit status 1
total build time: 0:00:08.229944, total build time today: 0:03:03.907742
```

Reply:
Hmm that's just because `unsafeToTensorShare` doesn't take a const argument. We should have a non-const overload :confused: